商品削除機能

aigis1 opened the pull request:
## What
商品削除機能の実装
## Why
出品者にのみ表示される削除ボタンを押すことで出品された商品を削除できるように

商品を削除した後にリダイレクトで飛ばされたトップページで商品を確認する様子
https://gyazo.com/e16f94fcb220bab68e3bc1416bf151d1

aigis1 committed:
destroyアクションへ処理の記述とリンクの設定

NaturalRadio commented on app/controllers/items_controller.rb:
```diff
@@ -38,6 +38,9 @@ def update
 
 
   def destroy
+    item = Item.find(params[:id])
```
```
item = Item.find(params[:id])
```
こちらbefore_action :set_itemにdestroyアクションを追加頂ければ記述を減らすことができますので修正してみましょう！

NaturalRadio commented on app/controllers/items_controller.rb:
```diff
@@ -38,6 +38,9 @@ def update
 
 
   def destroy
```
こちら、viewの条件と同様に「出品したユーザーでないと削除できない」という条件分岐を記述しておきましょう。

NaturalRadio reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

aigis1 committed:
destroyアクションにset_itemを追加、ユーザーによる条件分岐を追加

aigis1 commented:
## What
destroyアクションにset_itemを追加、ユーザーによる条件分岐を追加
## Why
記述の簡略化と削除機能を出品者以外に使えないようにするため

chirikuzu commented on app/controllers/items_controller.rb:
```diff
@@ -1,6 +1,6 @@
 class ItemsController < ApplicationController
   before_action :authenticate_user!, only: [:new, :edit]
-  before_action :set_item, only: [:show, :edit]
+  before_action :set_item, only: [:show, :edit, :destroy]
```
こちらの`set_item`ですが、`update`アクションにも適用させてあげることで`update`アクション内の`@item = Item.find(params[:id])`の記述を省略することが可能となります。
リファクタリングとしてぜひ修正してみましょう。

chirikuzu commented on app/controllers/items_controller.rb:
```diff
@@ -38,6 +38,12 @@ def update
 
 
   def destroy
+    if current_user == @item.user
+      @item.destroy
+      redirect_to root_path
+    else
+      move_to_index
+    end
```
こちらの記述ですが大変綺麗に修正できていますね。

`move_to_index`メソッドをprivate以下に定義していただいている場合、editアクションと同様にdestroyアクションに対してもbefore_actionとしてそちらを適用させていただいても良いかと考えられます。
destroyアクション内の記述が簡略化され、コードの可読性や保守性が向上するためです。

```ruby
# 4行目
before_action :move_to_index, only: [:edit] # ここに:destroyを追加する

def destroy
  @item.destroy
  redirect_to root_path
end
```

chirikuzu reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします。

aigis1 committed:
destroy, updateアクション内の記述を最適化

set_itemの対象にupdateアクションを追加し、記述を簡略化
destroyアクションのif文内で行っていたmove_to_indexをbefore_actionでの処理に変更した

aigis1 commented:
## What
destroy,updateアクション内の記述を最適化
## Why
記述と処理の簡略化

mizuki-iio commented on app/controllers/items_controller.rb:
```diff
@@ -38,6 +37,10 @@ def update
 
 
   def destroy
+    if current_user == @item.user
```
この条件はbefore_action :move_to_indexで実装しているため、ifでの条件分岐は不要になるかと思います！

mizuki-iio reviewed (commented):
レビュー依頼ありがとうございます！
1点コメントいたしましたので、ご確認をお願いいたします！

aigis1 committed:
不要な条件分岐の削除

aigis1 commented:
## What
if文の削除
## Why
before_actionで既に分岐を定義しているため不要であったから

chirikuzu reviewed (commented):
レビュー依頼ありがとうございます。
LGTMとさせていただきます。実装お疲れ様でした。
引き続き実装頑張っていきましょう。

aigis1 commented:
ありがとうございました
